File: src/session/filesystemWatcher.ts
import { filesystemService } from './filesystem'
import SessionConnection from "./sessionConnection"

export enum FilesystemOperation {
  Create = 'Create',
	Write = 'Write',
	Remove = 'Remove',
	Rename = 'Rename',
	Chmod = 'Chmod',
}

export interface FilesystemEvent {
  path: string
  operation: FilesystemOperation
  // Unix epoch in nanoseconds
  timestamp: number
  isDiretory: boolean
}

export type FilesystemEventListener = (event: FilesystemEvent) => void

class FilesystemWatcher {
  // Listeners to filesystem events.
  // Users of the this class can add their listeners to filesystem events
  // via `this.addEventListeners`
  private listeners: Set<FilesystemEventListener>
  private rpcSubscriptionID?: string

  constructor(
    private sessConn: SessionConnection,
    private path: string,
  ) {
    this.listeners = new Set<FilesystemEventListener>()
  }

  // Starts watching the path that was passed to the contructor
  async start() {
    // Already started.
    if (this.rpcSubscriptionID) return

    this.handleFilesystemEvents = this.handleFilesystemEvents.bind(this)

    this.rpcSubscriptionID = await this.sessConn.subscribe(
      filesystemService,
      this.handleFilesystemEvents,
      'watch',
      this.path,
    )
  }

  // Stops watching the path and removes all listeners.
  async stop() {
    if (this.rpcSubscriptionID) {
      await this.sessConn.unsubscribe(this.rpcSubscriptionID)
    }
    this.listeners.clear()
  }

  private handleFilesystemEvents(fsChange: FilesystemEvent) {
    this.listeners.forEach(l => {
      l(fsChange)
    })
  }

  addEventListener(l: FilesystemEventListener) {
    this.listeners.add(s)
    return {
      remove: () => {
        this.listeners.delete(s)
      }
    }
  }
}

export default FilesystemWatcher

File: packages/python-sdk/e2b/session/filesystem_watcher.py
from enum import Enum
from typing import Callable, Set, Any, Awaitable, Optional
from pydantic import BaseModel

from e2b.session.session_connection import SessionConnection
from e2b.session.exception import FilesystemException
from e2b.session.session_rpc import RpcException


class FilesystemOperation(str, Enum):
    Create = "Create"
    Write = "Write"
    Remove = "Remove"
    Rename = "Rename"
    Chmod = "Chmod"


class FilesystemEvent(BaseModel):
    path: str
    name: str
    operation: FilesystemOperation
    timestamp: int
    """
    Unix epoch in nanoseconds
    """
    is_dir: bool


class FilesystemWatcher:
    @property
    def path(self) -> str:
        """
        The path to watch.
        """
        return self._path

    def __init__(
        self,
        connection: SessionConnection,
        path: str,
        service_name: str,
    ):
        self._connection = connection
        self._path = path
        self._service_name = service_name
        self._unsubscribe: Optional[Callable[[], Awaitable[Any]]] = None
        self._listeners: Set[Callable[[FilesystemEvent], Any]] = set()

    async def start(self) -> None:
        """
        Starts the filesystem watcher.
        """
        if self._unsubscribe:
            return

        try:
            self._unsubscribe = await self._connection._subscribe(
                self._service_name,
                self._handle_filesystem_events,
                "watchDir",
                self.path,
            )
        except RpcException as e:
            raise FilesystemException(e.message) from e

    async def stop(self) -> None:
        """
        Stops the filesystem watcher.
        """
        self._listeners.clear()
        if self._unsubscribe:
            try:
                await self._unsubscribe()
                self._unsubscribe = None
            except RpcException as e:
                raise FilesystemException(e.message) from e

    def add_event_listener(self, listener: Callable[[FilesystemEvent], Any]):
        """
        Adds a listener for filesystem events.

        :param listener: a listener to add

        :return: a function that removes the listener
        """
        self._listeners.add(listener)

        def delete_listener() -> None:
            self._listeners.remove(listener)

        return delete_listener

    def _handle_filesystem_events(self, event: FilesystemEvent) -> None:
        for listener in self._listeners:
            listener(event)
